Add track installation instructions

codingthat opened the pull request.

codingthat committed:
Add track installation instructions

meatball133 commented on docs/INSTALLATION.md:
```diff
@@ -1,15 +1,15 @@
 # Installation
 
-<!-- TODO: write document
+## Installing Godot Engine
```
The idea here is that these should be instructions to setup the language for local testing. If it is meant that users should be using docker for testing then these docs should describe how to setup that for different os.

codingthat replied:
I'm not sure it's about testing, per se — there's both [docs/TESTS.md](https://exercism.org/docs/building/tracks/docs#h-file-tests-md) and [exercises/shared/tests.md](https://exercism.org/docs/building/tracks/shared-files#h-file-tests-md) for that.  The [spec and example for this docs/INSTALLATION.md](https://exercism.org/docs/building/tracks/docs#h-file-installation-md) seem to point more to just installing Godot itself, no?

meatball133 replied:
It is supposed to tell you how to install so you can test the tests locally.

I mean in most cases, say Python. You would install Python and then you could install the library needed and then simply run `pytest` and you would get the output. And some languages doesn't need any dependencies at all and have everything built in. So if you need like a bash script for it to work. Then these instructions should tell you how to install that script for example.

codingthat replied:
OK, fair enough — but then what goes in `TESTS.md` for Godot?  Should it just point back to `INSTALLATION.md` because there's nothing else to say like other tracks might (skipped tests, etc.)?

meatball133 replied:
So the difference between installation and tests is that installation says what/how you install to tests. But tests say how you test. So for python would that be something like:

INSTALLATION.md:

- How to install python
- How to install libraries

TESTS.md:

How to test e.g:

Find the folder with the exercise then `cd` into that folder

```
cd path/exercise
```

Then you have to call `pytest`.

codingthat replied:
@meatball133 when you have a moment, let me know if this part is OK.  Thanks!! :)

codingthat replied:
@meatball133 ?  :)

codingthat committed:
Inline refs and single-sentence-per-line

codingthat committed:
Include test setup instructions

meatball133 commented on docs/INSTALLATION.md:
````diff
@@ -1,15 +1,53 @@
+sudo mv gdscript-test-runner/ /opt/exercism/gdscript/test-runner/
+```
+
+### Step 2: Downloading the single-exercise test runner script
````
This part you actually don't have to include at all, since the user is expected to follow the general exercism instructions

codingthat replied:
I'm confused — these aren't part of the general exercism instructions...

meatball133 replied:
Wait my bad, didn't read to carefully, yes this should be here. You are right

meatball133 commented on docs/INSTALLATION.md:
```diff
@@ -1,15 +1,53 @@
+To build and test GDScript scripts for Exercism, Godot will be run in "headless" mode from the CLI.
+These instructions currently require Linux and bash.
+
+### Step 1: Installing the overall test runner infrastructure
```
But yes, this part should be here

codingthat commented on docs/INSTALLATION.md:
```diff
@@ -1,15 +1,53 @@
+Assuming you have the `exercism` tool set up and have downloaded at least one GDScript exercise, it should have created an `exercism/gdscript` folder to house the exercises.
+Save [the test runner][test-local-gdscript-solution] in this folder and mark the file executable.
+
+## Running tests
```
Lines 32-33 are still installation-related, though.  You need to do this just once, to be able to test any exercise after that.

We can duplicate it over in TESTS.md if you want, but, without the previous step it won't work.  So, in TESTS.md I can link to step 1 and duplicate step 2, if you want (but I'm not clear why).

codingthat replied:
Like, Step 1 is needed for students _and_ track maintainers.  Step 2 is an installation step that only students need.

meatball133 replied:
I only marked line 35, but it seemingly also shows the lines above. But yes, 32-33 should be here

BNAndras commented on docs/INSTALLATION.md:
```diff
@@ -1,15 +1,53 @@
 
-  This document should describe what the student needs to install
-  to allow working on the track on their local system using the CLI.
+The GDScript track is built on top of [Godot Engine][godot-engine], which runs on Windows, Linux, macOS, and Android.
```
This implies a student can solve exercises locally on those platforms, but we're later saying you need to run the test runner which apparently is Linux-only. So we should be clear upfront about this perhaps with an Exercism caution block so the student doesn't get frustrated after installing Godot locally.

codingthat replied:
Great point!  @meatball133 mentioned the possibility of providing a PowerShell equivalent to the local test runner, but since it depends on run.sh from the test runner repo, we'd need a PS equivalent there as well.  (Is it OK if I were to add a PowerShell todo to the GitHub issue tracker?)

Anyway, I will fix the wording here soon.

codingthat replied:
Done, cf. https://github.com/exercism/gdscript/pull/97/commits/107e96190ff28fd802854eeb2d2b78567c4f6869

BNAndras commented on docs/INSTALLATION.md:
```diff
@@ -1,15 +1,53 @@
+You'll need [Godot installed][installation] correctly to use the test runner.
+
+To build and test GDScript scripts for Exercism, Godot will be run in "headless" mode from the CLI.
+These instructions currently require Linux and bash.
```
Are we sure this requires just Linux? My Mac has /opt available and it can run Linux-based Docker images.

codingthat replied:
I have no Mac to test on, but if it can be verified to work there we can change it.  The local test runner described in this doc, though, doesn't use Docker (it relies on run.sh, which runs Godot directly), so you'd have to install Godot.

BNAndras replied:
Ah, we might want to call the local test runner something else. Within Exercism, a test runner is almost always the online one. That'll certainly cause some confusion down the line for example if a student is asking about running the local tests and we all think they refer to the online test runner.

BNAndras replied:
I have access to a M1 Mac so I can check this script out later this week.

codingthat replied:
I see some places I've called it simply "test runner" and I could make it "local test runner" to be unambiguous.  But if you think that will still be confused with the online one despite "local" everywhere, I'm open to suggestions.  I see other tracks don't generally need one because the CLI syntax for testing is a bit more standardized than in the Godot ecosphere.

Maybe "local solution checker"?

codingthat replied:
@BNAndras what do you think?

BNAndras replied:
Works fine for me on a Mac.

```bash
(base) ➜  darts git:(main) ✗ ./test-local-gdscript-solution.sh
Godot Engine v4.7.stable.official.5b4e0cb0f - https://godotengine.org

-> test_missed_target
-> test_on_the_outer_circle
-> test_on_the_middle_circle
-> test_on_the_inner_circle
-> test_exactly_on_center
-> test_near_the_center
-> test_just_within_the_inner_circle
-> test_just_outside_the_inner_circle
-> test_just_within_the_middle_circle
-> test_just_outside_the_middle_circle
-> test_just_within_the_outer_circle
-> test_just_outside_the_outer_circle
-> test_asymmetric_position

Exercise passed!  Nicely done!
```

As for the terms, this gets complicated because normally what gets written in these cases is a testing library that the online test runner hooks into but still can be run locally separately from the test runner. However, in this specific case we're using a script to run the test runner locally so it's the same experience  whether you're in the online editor or working locally. I guess it really is just the test runner so maybe the best we can do here is consistently specify local test runner vs. online test runner.

BNAndras commented on docs/INSTALLATION.md:
```diff
@@ -1,15 +1,53 @@
+
+### Step 2: Downloading the single-exercise test runner script
+
+Assuming you have the `exercism` tool set up and have downloaded at least one GDScript exercise, it should have created an `exercism/gdscript` folder to house the exercises.
```
The `exercism` tool hasn't been previously mentioned so we can't assume students have done any of this. We should link to the the CLI installation guide and then provide a description of how to find that `exercism/gdscript` folder in case the student doesn't know where to find it.

codingthat replied:
Done, cf. https://github.com/exercism/gdscript/pull/97/commits/107e96190ff28fd802854eeb2d2b78567c4f6869 and https://github.com/exercism/gdscript/pull/97/commits/c82456fc08d9632c6e65fceac931951abe364d89 .

BNAndras commented on docs/INSTALLATION.md:
```diff
@@ -1,15 +1,53 @@
+
+## Using an IDE
+
+For syntax checking of GDScript files from, for example, [Visual Studio Code][vscode], you'll need to keep the Godot GUI running with a project open in order to activate the debug server, and an extension like [godot-tools][godot-tools] installed and activated.
```
This could use some information about how one might activate the debug server. 

Also, is it checking/verifying the syntax or highlighting it? The godot-tools page describes syntax highlighting, and those are different things.

codingthat replied:
I guess it just needs rewording: Having Godot running (with a project open, as mentioned) _is_ how one activates the debug server.  There's no additional step.

It does both, and more, I think, depending on what checking/verifying is considered to cover.  For example, if I add a duplicate `return` statement, it does this:

<img width="1478" height="269" alt="Tooltip warning about unreachable code" src="https://github.com/user-attachments/assets/479c9c6c-f719-4d99-999c-d7bf0d4816ab" />

BNAndras replied:
Yeah so it's doing some sort of static analysis then. That's fine.

We could just link to the godot-tools docs and have the students follow up the official setup there including keeping Godot running. That keeps things focused on installing and getting Godot set up for Exercism use.

codingthat replied:
Done, cf. https://github.com/exercism/gdscript/commit/107e96190ff28fd802854eeb2d2b78567c4f6869

codingthat replied:
Ah, sorry, I didn't see your message before writing that.

We could, but actually their official documentation seems incorrect on that point:

> Make sure the Godot editor is running
> Make sure to open the project in the Godot editor first. If you opened the editor after opening VS Code, you can click the Retry button in the bottom-right corner in VS Code.

The specific project doesn't need to be open.  Any project will do.  No need to keep the Godot editor paralleling your exercises—just open something and forget about it.  So maybe leaving our note (now that I've improved the wording, that is) will be helpful to learners?

codingthat replied:
@BNAndras What do you think, is it OK to leave the note after all, since it's more accurate than the official docs?  I'm still seeing the wording I quoted e.g. at https://marketplace.visualstudio.com/items?itemName=geequlim.godot-tools

BNAndras replied:
Seems fine. I might tighten up the prose a little like this:
```
To get GDScript syntax highlighting inside an editor like [Visual Studio Code][vscode], you'll need an extension such as [godot-tools][godot-tools] installed plus a running debug server.
Godot starts this debug server automatically whenever a project is open in its own editor and keeps it running until the project or Godot itself is closed.
````

codingthat replied:
Done in [68b0f4d](https://github.com/exercism/gdscript/pull/97/commits/68b0f4ded904fb528a870685d113a26ea583223d).  Thanks!

codingthat committed:
Add clarifications:

- Linux requirement for local test runner
- Syntax highlighting "etc."
- Requirements for keeping a debug server running

codingthat committed:
Merge branch 'exercism:main' into add-installation

codingthat committed:
Fix exercism workspace location

BNAndras commented on docs/INSTALLATION.md:
```diff
@@ -1,15 +1,53 @@
+### Step 2: Downloading the single-exercise test runner script
+
+Assuming you have the `exercism` tool set up and have downloaded at least one GDScript exercise, it should have created an `exercism/gdscript` folder to house the exercises.
+Save [the test runner][test-local-gdscript-solution] in this folder and mark the file executable.
```
If we're using a custom testing setup, I strongly recommend that this file be provided to students during the initial download for each exercise. That's what we do on other tracks with custom testing libraries like Red and Scheme. Downloading it afterwards as a separate step runs the risk of confusing students. If I'm downloading the files to work offline, my general expectation would be I can start working with minimal fuss.

codingthat committed:
Tighten up the prose re: "Using an IDE"

Per suggestion by @BNAndras
